### PR TITLE
1.4

### PR DESCRIPTION
**Pull request by rmehlinger**
Remove dead code
Consistently use `all()` instead of `raw()`
nullcheck map event handlers
fixed `ObsMap.has`
added `ObsMap.keys` and `ObsMap.values` utility functions

Finally, fleshed out `ObsArray` and `SrcArray` with reactive equivalents of most of the member methods of the native JavaScript Array class. In addition added some other useful methods that seemed like a good idea to have around:

`transform`: a replacement for the `rx.cellToArray rx.bind => @all()...` dance, which also avoids the creation of an unnecessary intermediate cell.
`move` and `swap`: Highly useful for UI work, e.g. with drag-and-drops, and more complicated than they seem at first glance.
`first` and `last`: Some simple syntactic sugar, a l'underscore.

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "reactive-coffee",
-  "version": "1.3.2",
+  "version": "1.3.3",
```
Comment (yang) [inline bower.json]: This already is at 1.3.3....

File: src/reactive.coffee [hunk 2]
```diff
@@ -432,11 +432,11 @@ rxFactory = (_, $) ->
       @onChange = new Ev() # {key: [old, new]...}
     get: (key) ->
       recorder.sub (target) => rx.autoSub @onAdd, (additions) ->
-        target.refresh() if key of additions
+        target.refresh() if key of (additions ? {})
```
Comment (yang) [inline src/reactive.coffee:475]: These should never be null

**Review by yang — CHANGES_REQUESTED** [review]
I'd abandon this commit

File: src/reactive.coffee [hunk 3]
```diff
@@ -10,14 +10,6 @@ rxFactory = (_, $) ->
     delete x[k]
     v
 
-  nthWhere = (xs, n, f) ->
```
Comment (yang) [inline src/reactive.coffee:13]: I would split these unrelated changes out to minimize noise

File: package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "reactive-coffee",
-  "version": "1.3.3",
+  "version": "1.4.0",
```
Comment (yang) [inline package.json:3]: Why cut a new release?

Also, release bumps should always be in their own commit, not bundled with other code changes.

File: src/reactive.coffee [hunk 5]
```diff
@@ -135,7 +127,7 @@ rxFactory = (_, $) ->
     sub: (sub) ->
       if @stack.length > 0 and not @isIgnoring
         topCell = _(@stack).last()
-        handle = sub(topCell)
+        sub(topCell)
```
Comment (yang) [inline src/reactive.coffee:130]: Ditto as above

File: src/reactive.coffee [hunk 6]
```diff
@@ -298,16 +290,26 @@ rxFactory = (_, $) ->
...
       rx.autoSub @onChangeCells, ([index, removed, added]) =>
         for cell in ys.cells[index...index + removed.length]
           cell.disconnect()
-        newCells =
-          added.map (item) ->
-            cell = bind -> f(item.get())
+        newCells = added.map (item) -> bind -> f(item.get())
         ys.realSpliceCells(index, removed.length, newCells)
       ys
+    transform: (f, diff) -> new DepArray (=> f @all()), diff
```
Comment (yang) [inline src/reactive.coffee:301]: These all need to be deeply tested (specifying the behavior of incremental changes on the transforms)

File: src/reactive.coffee [hunk 7]
```diff
@@ -343,13 +343,62 @@ rxFactory = (_, $) ->
     splice: (index, count, additions...) -> @spliceArray(index, count, additions)
     insert: (x, index) -> @splice(index, 0, x)
     remove: (x) ->
-      i = _(@raw()).indexOf(x)
+      i = _(rx.snap => @all()).indexOf(x)
```
Comment (yang) [inline src/reactive.coffee:346]: Why move to all()? Slower, and the original use of `raw()` was safe / non-mutating

Same various points below

Also really need to break these out into separate commits

File: src/reactive.coffee [hunk 8]
```diff
@@ -343,13 +343,62 @@ rxFactory = (_, $) ->
...
       @removeAt(i) if i >= 0
-    removeAt: (index) -> @splice(index, 1)
+    removeAll: (x) -> rx.transaction =>
```
Comment (yang) [inline src/reactive.coffee:348]: Same with testing all of these

File: src/reactive.coffee [hunk 9]
```diff
@@ -439,12 +479,18 @@ rxFactory = (_, $) ->
         target.refresh() if key of (removals ? {})
       @x[key]
     has: (key) ->
-      @x[key]?
+      recorder.sub (target) => rx.autoSub @onAdd, (additions) ->
```
Comment (yang) [inline src/reactive.coffee:482]: Great catch

**Review by yang — CHANGES_REQUESTED** [review]

File: src/reactive.coffee (hunk 7 above)
Comment (rmehlinger) [reply src/reactive.coffee:346]: `raw` already calls `all`. I'm not sure who made that change, but there seems to be no reason for its continued existence.

File: src/reactive.coffee (hunk 6 above)
Comment (rmehlinger) [reply src/reactive.coffee:301]: Yup, planning on giving them the same treatment as the `SrcArray` methods. By incremental changes, I assume you mean things like adding and removing one at a time?

**Pull request closed without merging**